Fixes the wrong use of k8s client-go

Nino-K opened the pull request:
Previously we were using the client incorrectly by not passing the context, this was causing internal network errors in the client due to the API server not being ready as you can see in the below logs:
```
2024/05/03 09:03:12 [DEBUG]   list of services from core v1----> []
W0503 09:03:12.649264     255 reflector.go:424] pkg/mod/k8s.io/client-go@v0.26.2/tools/cache/reflector.go:169: failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: network is unreachable
2024/05/03 09:03:12 [DEBUG]   watching kubernetes services
2024/05/03 09:03:12 [DEBUG]   inside state watching
2024/05/03 09:03:12 [DEBUG]   kubernetes: error watching [error=failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: network is unreachable]
2024/05/03 09:03:12 [ERROR]   kubernetes: unexpected error watching [error=failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: network is unreachable]
W0503 09:03:14.012174     255 reflector.go:424] pkg/mod/k8s.io/client-go@v0.26.2/tools/cache/reflector.go:169: failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: connection refused
2024/05/03 09:03:14 [DEBUG]   kubernetes: error watching [error=failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: connection refused]
2024/05/03 09:03:14 [DEBUG]   kubernetes: got error, rolling back [error=failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 127.0.0.1:6443: connect: connection refused]
2024/05/03 09:03:15 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
```
 However, we continued and blocked the event channel. 

With the new approach, we can determine when the API server is ready, this way we return a FATAL error and the process exists. This allows the init process to retry again until the API server becomes ready.

The below logs are for the the new approach:

```
2024/05/03 09:29:58 [INFO]    Starting Rancher Desktop Agent in [AdminInstall=true] mode
2024/05/03 09:29:58 [DEBUG]   kubernetes: loaded kubeconfig /etc/rancher/k3s/k3s.yaml
2024/05/03 09:29:58 [FATAL]   error watching services: error listing services: Get "https://127.0.0.1:6443/api/v1/services": dial tcp 127.0.0.1:6443: connect: network is unreachable
2024/05/03 09:30:03 [INFO]    Starting Rancher Desktop Agent in [AdminInstall=true] mode
2024/05/03 09:30:03 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:04 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:05 [DEBUG]   kubernetes: failed to read kubeconfig [error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory][config-path=/etc/rancher/k3s/k3s.yaml]
2024/05/03 09:30:06 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:07 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:08 [DEBUG]   checking if container engine API is running at /var/run/docker.sock
2024/05/03 09:30:08 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:09 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:10 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:11 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:12 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:13 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:14 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:15 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:16 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:17 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:18 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:19 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:20 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:21 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:22 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:23 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:24 [DEBUG]   kubernetes: failed to read kubeconfig [config-path=/etc/rancher/k3s/k3s.yaml][error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory]
2024/05/03 09:30:25 [DEBUG]   kubernetes: failed to read kubeconfig [error=could not load Kubernetes client config from /etc/rancher/k3s/k3s.yaml: stat /etc/rancher/k3s/k3s.yaml: no such file or directory][config-path=/etc/rancher/k3s/k3s.yaml]
2024/05/03 09:30:26 [DEBUG]   kubernetes: loaded kubeconfig /etc/rancher/k3s/k3s.yaml
2024/05/03 09:30:26 [ERROR]   context cancellation: context canceled
2024/05/03 09:30:26 [FATAL]   error watching services: error listing services: apiserver not ready
2024/05/03 09:30:31 [INFO]    Starting Rancher Desktop Agent in [AdminInstall=true] mode
2024/05/03 09:30:31 [DEBUG]   kubernetes: loaded kubeconfig /etc/rancher/k3s/k3s.yaml
```

Fixes: https://github.com/rancher-sandbox/rancher-desktop/issues/6809

Nino-K committed:
Fixes the wrong use of k8s client-go

Previously we were using the client incorrectly by not passing the
context, this was causing internal network error in the client due to
the API server not being ready. However, we continued and blocked on the
event channel. With the new approach, we can determine when the API server
is ready, this way we return a FATAL error and the process exists. This
allows the init process to retry again until the API server becomes
ready.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on pkg/kube/servicewatcher.go:
```diff
@@ -48,14 +48,17 @@ func watchServices(ctx context.Context, client *kubernetes.Clientset) (<-chan ev
 	serviceInformer := informerFactory.Core().V1().Services()
 	sharedInformer := serviceInformer.Informer()
-	_, _ = sharedInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	_, _ = serviceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Is there a reason we're no longer using `sharedInformer` here? It seems to be the same thing here.

mook-as commented on pkg/kube/servicewatcher.go:
```diff
@@ -48,14 +48,17 @@ func watchServices(ctx context.Context, client *kubernetes.Clientset) (<-chan ev
-	_, _ = sharedInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	_, _ = serviceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
+			log.Debugf("Service Informer: Add func called with: %+v", obj)
```
Not that it matters, but these logs should be trace?

Nino-K replied:
AFAIK for the trace logging to work it needs to be handled from the [main](https://github.com/rancher-sandbox/rancher-desktop-agent/blob/078aaed77ef16f0d62357340a9eb27df309df048/cmd/rancher-desktop-guestagent/main.go#L94). I don't think just changing these to trace would work 🤔 
Also, I don't recall if we handle trace logging at all in RD anywhere.

mook-as replied:
Yeah, trace would mean it's basically off and needs a code change to turn on. We don't really want to log everything in release builds (even with debug turned on).

mook-as commented on pkg/kube/servicewatcher.go:
```diff
@@ -101,15 +104,16 @@ func watchServices(ctx context.Context, client *kubernetes.Clientset) (<-chan ev
 	if err != nil {
 		return nil, nil, fmt.Errorf("error listing services: %w", err)
 	}
+	log.Debugf("coreV1 services list :%+v", services.Items)
```
Same here — trace?

mook-as commented on pkg/kube/servicewatcher.go:
```diff
@@ -101,15 +104,16 @@ func watchServices(ctx context.Context, client *kubernetes.Clientset) (<-chan ev
 	informerFactory.Start(ctx.Done())
 
-	services, err := serviceInformer.Lister().List(labels.Everything())
+	services, err := client.CoreV1().Services(corev1.NamespaceAll).List(ctx, v1.ListOptions{})
```
Discussed with Nino: This manages to return an error (good!), but we don't handle it (bad!) and exit the whole process. At least `init` restarts us, but we should avoid needing to do that.

Nino-K committed:
Handle the return error from listing services

This will prevent the process to exit with a FATAL error
which prevents the init process restarting guest agent again.
The error is now handled internally withing the process.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on pkg/kube/watcher.go:
```diff
@@ -120,8 +122,12 @@ func WatchForServices(
+				case isTimeout(err):
+				case errors.Is(err, unix.ENETUNREACH):
+				case errors.Is(err, unix.ECONNREFUSED):
+				case apierrors.IsInternalError(err):
```
That's for _all_ internal errors; I guess that's fine?

Nino-K committed:
Fixes the lint error for magic number

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on pkg/kube/watcher.go:
```diff
@@ -258,6 +267,10 @@ func isTimeout(err error) bool {
 }
 
+func isAPINotReady(err error) bool {
+	return strings.Contains(err.Error(), "apiserver not ready")
```
It would be good to have a comment about where this comes from (k3s, over HTTP so we can't test for a specific error object). Ideally we also check that `err` is of the right type here, but I guess that can be optional.

mook-as commented on pkg/kube/watcher.go:
```diff
@@ -120,14 +122,21 @@ func WatchForServices(
-					// If it's a time out, the server may not be running yet
+				switch {
+				case isTimeout(err):
+					fallthrough
```
This might be clearer as:
```golang
switch {
default:
  return err
case isTimeout(err):
case errors.Is(err, unix.ENETUNREACH):
case errors.Is(err, unix.ECONNREFUSED):
case isAPINotReady(err):
}
// Sleep and continue for all the expected errors
time.Sleep(time.Second)
continue
```

mook-as commented on cmd/rancher-desktop-guestagent/main.go:
```diff
@@ -181,7 +181,7 @@ func main() {
 
 			if k8sServiceListenerIP == nil || !(k8sServiceListenerIP.Equal(net.IPv4zero) ||
-				k8sServiceListenerIP.Equal(net.IPv4(127, 0, 0, 1))) { //nolint:gomnd // IPv4 addr localhost
+				k8sServiceListenerIP.Equal(net.IPv4(127, 0, 0, 1))) { //nolint:gomnd,mnd // IPv4 addr localhost
```
It might be better to update the lint config to allow magic numbers for `net.IPv4` instead.

Nino-K committed:
White list net.IPv4 function for gomnd linter

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Improve readability of WatchForService function

Signed-off-by: Nino Kodabande <nkodabande@suse.com>
